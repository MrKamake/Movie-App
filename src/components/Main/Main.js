import React, { Component } from 'react';
import './Main.scss';
import { Link } from 'react-router-dom';
import { MOVIE_IMAGE_URL } from '../../constants/constant';
import Noimage from '../../image/Noimage.png';
import debounce from 'lodash/debounce';

class Main extends Component {
  constructor(props) {
    super(props);

    this.state = {
      pageNumber: 1
    };
    this.scrollEvent = debounce(this._scrollPage, 200);
  }

  componentDidMount() {
    window.addEventListener('scroll', this.scrollEvent);
    const { onLoad } = this.props;
    const { pageNumber } = this.state;
    onLoad(pageNumber);
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.scrollEvent);
  }

  _scrollPage = () => {
    const { onLoad } = this.props;
    const { pageNumber } = this.state;

    let scrollHeight = Math.max(
      document.documentElement.scrollHeight,
      document.body.scrollHeight
    );
    let scrollTop = Math.max(
      document.documentElement.scrollTop,
      document.body.scrollTop
    );
    let clientHeight = document.documentElement.clientHeight;

    if (scrollHeight - scrollTop - clientHeight <= 200) {
      this.setState(
        {
          pageNumber: pageNumber + 1
        },
        onLoad(pageNumber)
      );
    }
  };

  render() {
    const { movieList, genreList } = this.props;

    return (
      <div className="Main">
        {movieList.map((movie, index) => (
          <Link to={`/movie/${movie.id}`} className="movie-element" key={index}>
            {movie.poster_path ? (
              <img
                src={MOVIE_IMAGE_URL + movie.poster_path}
                alt={movie.title}
              />
            ) : (
              <img src={Noimage} alt="Noimage" />
            )}
            <div className="movie-description">
              <div className="movie-title">{movie.title}</div>
              <div className="movie-rating">{movie.vote_average}</div>
              <div className="movie-date">{movie.release_date}</div>
              <div className="movie-genre-list">
                {genreList.map((genre, index) => (
                  <div className="movie-genre-wrapper" key={index}>
                    {movie.genre_ids.map((movieGenre, idx) => (
                      <>
                        {genre.id === movieGenre && (
                          <div className="movie-genre" key={idx}>
                            {genre.name}
                          </div>
                        )}
                      </>
                    ))}
                  </div>
                ))}
              </div>
              <p className="movie-overview">{movie.overview}</p>
            </div>
          </Link>
        ))}
      </div>
    );
  }
}

export default Main;
