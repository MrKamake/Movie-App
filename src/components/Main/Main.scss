$ratingColor: gold;

.Main {
  text-align: center;
  color: white;
  .movie-element {
    display: inline-block;
    margin: 40px 20px;
    width: 640px;
    text-align: left;
    color: white;
    img {
      width: 200px;
    }
    .movie-description {
      display: inline-block;
      padding-left: 10px;
      .movie-title {
        width: 400px;
        padding-bottom: 50px;
        font-size: 1.5rem;
        font-weight: bold;
        overflow: hidden;
      }
      .movie-rating {
        display: inline-block;
        margin: 0 10px 10px 0;
        padding: 2px 6px;
        border-radius: 10px;
        color: black;
        background: $ratingColor;
      }
      .movie-date {
        display: inline-block;
      }
      .movie-genre-list {
        padding: 0 1px;
        .movie-genre-wrapper {
          display: inline-block;
          .movie-genre {
            margin: 3px;
            padding: 2px 5px;
            border-radius: 4px;
            background: gray;
          }
        }
      }
      .movie-overview {
        width: 400px;
        height: 50px;
        font-size: 0.9rem;
        overflow: hidden;
      }
    }
  }
}
