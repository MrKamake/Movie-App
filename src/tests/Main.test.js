import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Main from '../components/Main/Main';

configure({ adapter: new Adapter() });

describe('<Main />', () => {
  let wrapper;

  beforeEach(() => {
    const onLoad = jest.fn();
    const getNextPage = jest.fn();
    const movieList = [
      {
        poster_path: '/asdfoo3.jpg',
        title: 'Life of vaco',
        vote_average: 4,
        release_date: 2019 - 12 - 11,
        genre_ids: [1, 2, 5],
        overview: 'askdfja2qbnelvpaoisfkw'
      }
    ];
    const genreList = [];
    wrapper = shallow(
      <Main
        onLoad={onLoad}
        getNextPage={getNextPage}
        movieList={movieList}
        genreList={genreList}
      />
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should contain element', () => {
    expect(
      wrapper.contains(<div className="movie-title">Life of vaco</div>)
    ).toBe(true);
    expect(
      wrapper.contains(<p className="movie-overview">askdfja2qbnelvpaoisfkw</p>)
    ).toBe(true);
  });
});
