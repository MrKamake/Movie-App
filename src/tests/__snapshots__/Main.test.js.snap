// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main /> matches snapshot 1`] = `
<div
  className="Main"
>
  <Link
    className="movie-element"
    key="0"
    to="/movie/undefined"
  >
    <img
      alt="Life of vaco"
      src="https://image.tmdb.org/t/p/w300_and_h450_bestv2/asdfoo3.jpg"
    />
    <div
      className="movie-description"
    >
      <div
        className="movie-title"
      >
        Life of vaco
      </div>
      <div
        className="movie-rating"
      >
        4
      </div>
      <div
        className="movie-date"
      >
        1996
      </div>
      <div
        className="movie-genre-list"
      />
      <p
        className="movie-overview"
      >
        askdfja2qbnelvpaoisfkw
      </p>
    </div>
  </Link>
</div>
`;
